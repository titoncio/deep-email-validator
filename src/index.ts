import { isEmail, parseEmail } from './regex/regex'
import { checkTypo } from './typo/typo'
import { getBestMx } from './dns/dns'
import { checkSMTP } from './smtp/smtp'
import { checkDisposable } from './disposable/disposable'
import { getOptions, ValidatorOptions } from './options/options'
import { OutputFormat, createOutput } from './output/output'
import './types'

export async function validate(emailOrOptions: string | ValidatorOptions): Promise<OutputFormat> {
  const options = getOptions(emailOrOptions)
  const emailRaw = options.email

  const regexResponse = parseEmail(emailRaw, {
    allowQuoted: options.allowQuoted,
    allowAngle: options.allowAngle,
    rejectSubaddressing: options.rejectSubaddressing,
  })
  if (options.validateRegex && 'error' in regexResponse) return createOutput('regex', regexResponse.error)
  // fallback to the naive domain extraction if the user specifically opted out of format validation
  const domain = 'domain' in regexResponse ? regexResponse.domain : emailRaw.split('@')[1]
  const email = 'effectiveAddr' in regexResponse ? regexResponse.effectiveAddr : emailRaw.trim()

  // prevent SMTP injection
  if (email.indexOf('\r') !== -1 || email.indexOf('\n') !== -1) {
    return createOutput('sanitization', 'Email cannot contain newlines')
  }

  if (options.validateTypo) {
    const typoResponse = await checkTypo(email, options.additionalTopLevelDomains)
    if (typoResponse) return createOutput('typo', typoResponse)
  }

  if (options.validateDisposable) {
    const disposableResponse = await checkDisposable(domain)
    if (disposableResponse) return createOutput('disposable', disposableResponse)
  }

  if (options.validateMx) {
    const mx = await getBestMx(domain)
    if (!mx) return createOutput('mx', 'MX record not found')
    if (options.validateSMTP) {
      return checkSMTP(options.sender, email, mx.exchange)
    }
  }

  return createOutput()
}

export default validate
