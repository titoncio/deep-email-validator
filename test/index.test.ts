import values from 'lodash/values'
import every from 'lodash/every'
import { validate } from '../src/index'

const elevenSeconds = 11 * 1000

describe('validation tests', () => {
  it('fails without sending data', async () => {
    const res = await validate({
      email: 'jubao@le.com',
      sender: 'jubao@le.com',
      validateTypo: false,
      validateDisposable: false,
    })
    expect(res.valid).toBe(false)
    expect(res.reason).toBe('smtp')
    expect(res.validators.smtp?.valid).toBe(false)
    expect(res).toMatchSnapshot()
  })
  it('fails with bad regex', async () => {
    const res = await validate('david.gmail.com')
    expect(res.valid).toBe(false)
    expect(res.reason).toBe('regex')
    expect(res.validators.regex?.valid).toBe(false)
    expect(res).toMatchSnapshot()
  })

  it('fails later with malformed regex', async () => {
    const res = await validate({
      email: 'dav id@gmail.com',
      validateRegex: false,
    })
    expect(res.valid).toBe(false)
    expect(res.reason).toBe('smtp')
    expect(res.validators.regex?.valid).toBe(true)
    expect(res.validators.smtp?.valid).toBe(false)
    expect(res).toMatchSnapshot()
  })

  it('fails with common typo', async () => {
    const res = await validate('david@gmaill.com')
    expect(res.valid).toBe(false)
    expect(res.reason).toBe('typo')
    expect(res.validators.typo?.valid).toBe(false)
    expect(res).toMatchSnapshot()
  })

  it('fails with disposable email', async () => {
    const res = await validate('david@temp-mail.org')
    expect(res.valid).toBe(false)
    expect(res.reason).toBe('disposable')
    expect(res.validators.disposable?.valid).toBe(false)
    expect(res).toMatchSnapshot()
  })

  it(
    'fails with bad dns',
    async () => {
      const res = await validate('xxx@yyy.zzz')
      expect(res.valid).toBe(false)
      expect(res.reason).toBe('mx')
      expect(res.validators.mx?.valid).toBe(false)
      expect(res).toMatchSnapshot()
    },
    elevenSeconds
  )

  it('fails with bad mailbox', async () => {
    const res = await validate('david@andco.life')
    expect(res.valid).toBe(false)
    expect(res.reason).toBe('smtp')
    expect(res.validators.smtp?.valid).toBe(false)
    expect(res).toMatchSnapshot()
  })

  it(
    'passes when we skip smtp validation',
    async () => {
      const res = await validate({
        email: 'david@andco.life',
        validateSMTP: false,
      })
      expect(res.valid).toBe(true)
      expect(every(values(res.validators), x => x && x.valid)).toBe(true)
      expect(res).toMatchSnapshot()
    },
    elevenSeconds
  )

  it(
    'passes when valid special char',
    async () => {
      const res = await validate('~@oftn.org')
      expect(res.valid).toBe(true)
      expect(every(values(res.validators), x => x && x.valid)).toBe(true)
      expect(res).toMatchSnapshot()
    },
    elevenSeconds
  )

  it(
    'passes when quoted address',
    async () => {
      const res = await validate({ email: '"this+is+my+personal+email+address@me.invalid"@gmail.com', validateRegex: true, validateMx: true, validateSMTP: false, allowQuoted: true })
      expect(res.valid).toBe(true)
      expect(every(values(res.validators), x => x && x.valid)).toBe(true)
      expect(res).toMatchSnapshot()
    }
  )

  it(
    'fails when quoted address is not allowed',
    async () => {
      const res = await validate({ email: '"this+is+my+personal+email+address@me.invalid"@gmail.com', validateRegex: true, validateMx: true, validateSMTP: false })
      expect(res.valid).toBe(false)
      expect(res.validators.regex?.valid).toBe(false)
    }
  )

  it(
    'passes when subaddressing is allowed',
    async () => {
      const res = await validate({ email: 'me+test@gmail.com', validateRegex: true, validateMx: true, validateSMTP: false })
      expect(res.valid).toBe(true)
      expect(every(values(res.validators), x => x && x.valid)).toBe(true)
      expect(res).toMatchSnapshot()
    }
  )


  it(
    'fails when subaddressing is not allowed',
    async () => {
      const res = await validate({
        email: 'me+test@gmail.com', validateRegex: true, validateMx: true, validateSMTP: false, rejectSubaddressing: true
      })
      expect(res.valid).toBe(false)
      expect(res.validators.regex?.valid).toBe(false)
      expect(res).toMatchSnapshot()
    }
  )

  it(
    'passes when angle address is allowed',
    async () => {
      const res = await validate({ email: 'Me <me@gmail.com>', validateRegex: true, validateMx: true, validateSMTP: false, allowAngle: true })
      expect(res.valid).toBe(true)
      expect(every(values(res.validators), x => x && x.valid)).toBe(true)
      expect(res).toMatchSnapshot()
    }
  )

  it(
    'fails when angle address is not allowed',
    async () => {
      const res = await validate({ email: 'Me <me@gmail.com>', validateRegex: true, validateMx: true, validateSMTP: false })
      expect(res.valid).toBe(false)
      expect(res.validators.regex?.valid).toBe(false)
      expect(res).toMatchSnapshot()
    }
  )

  it(
    'passes when valid wildcard',
    async () => {
      const res = await validate('info@davidalbertoadler.com')
      expect(res.valid).toBe(true)
      expect(every(values(res.validators), x => x && x.valid)).toBe(true)
      expect(res).toMatchSnapshot()
    },
    elevenSeconds
  )

  it(
    'passes with custom TLD',
    async () => {
      const res = await validate({
        email: 'info@utob.ir',
        validateSMTP: false,
        additionalTopLevelDomains: ['ir'],
      })
      expect(res.valid).toBe(true)
      expect(every(values(res.validators), x => x && x.valid)).toBe(true)
      expect(res).toMatchSnapshot()
    },
    elevenSeconds
  )
})
