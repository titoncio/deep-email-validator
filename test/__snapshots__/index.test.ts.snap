// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation tests fails later with malformed regex 1`] = `
Object {
  "reason": "smtp",
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "valid": true,
    },
    "mx": Object {
      "valid": true,
    },
    "regex": Object {
      "valid": true,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "reason": "SMTP Error: The mail address that you specified was not syntactically correct.",
      "valid": false,
    },
    "typo": Object {
      "valid": true,
    },
  },
}
`;

exports[`validation tests fails when angle address is not allowed 1`] = `
Object {
  "reason": "regex",
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "valid": false,
    },
    "mx": Object {
      "valid": false,
    },
    "regex": Object {
      "reason": "Angle address is not allowed",
      "valid": false,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "valid": false,
    },
    "typo": Object {
      "valid": false,
    },
  },
}
`;

exports[`validation tests fails when subaddressing is not allowed 1`] = `
Object {
  "reason": "regex",
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "valid": false,
    },
    "mx": Object {
      "valid": false,
    },
    "regex": Object {
      "reason": "Subaddressing is not allowed",
      "valid": false,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "valid": false,
    },
    "typo": Object {
      "valid": false,
    },
  },
}
`;

exports[`validation tests fails with bad dns 1`] = `
Object {
  "reason": "mx",
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "valid": true,
    },
    "mx": Object {
      "reason": "MX record not found",
      "valid": false,
    },
    "regex": Object {
      "valid": true,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "valid": false,
    },
    "typo": Object {
      "valid": true,
    },
  },
}
`;

exports[`validation tests fails with bad mailbox 1`] = `
Object {
  "reason": "smtp",
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "valid": true,
    },
    "mx": Object {
      "valid": true,
    },
    "regex": Object {
      "valid": true,
    },
    "smtp": Object {
      "reason": "Mailbox not found.",
      "valid": false,
    },
    "typo": Object {
      "valid": true,
    },
  },
}
`;

exports[`validation tests fails with bad regex 1`] = `
Object {
  "reason": "regex",
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "valid": false,
    },
    "mx": Object {
      "valid": false,
    },
    "regex": Object {
      "reason": "Email does not contain \\"@\\".",
      "valid": false,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "valid": false,
    },
    "typo": Object {
      "valid": false,
    },
  },
}
`;

exports[`validation tests fails with common typo 1`] = `
Object {
  "reason": "typo",
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "valid": false,
    },
    "mx": Object {
      "valid": false,
    },
    "regex": Object {
      "valid": true,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "valid": false,
    },
    "typo": Object {
      "reason": "Likely typo, suggested email: david@gmail.com",
      "valid": false,
    },
  },
}
`;

exports[`validation tests fails with disposable email 1`] = `
Object {
  "reason": "disposable",
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "reason": "Email was created using a disposable email service",
      "valid": false,
    },
    "mx": Object {
      "valid": false,
    },
    "regex": Object {
      "valid": true,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "valid": false,
    },
    "typo": Object {
      "valid": true,
    },
  },
}
`;

exports[`validation tests fails without sending data 1`] = `
Object {
  "reason": "smtp",
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "valid": true,
    },
    "mx": Object {
      "valid": true,
    },
    "regex": Object {
      "valid": true,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "reason": "SMTP Error: The requested action was not done. Some error occurmiles in the mail server.",
      "valid": false,
    },
    "typo": Object {
      "valid": true,
    },
  },
}
`;

exports[`validation tests passes when angle address is allowed 1`] = `
Object {
  "valid": true,
  "validators": Object {
    "disposable": Object {
      "valid": true,
    },
    "mx": Object {
      "valid": true,
    },
    "regex": Object {
      "valid": true,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "valid": true,
    },
    "typo": Object {
      "valid": true,
    },
  },
}
`;

exports[`validation tests passes when quoted address 1`] = `
Object {
  "valid": true,
  "validators": Object {
    "disposable": Object {
      "valid": true,
    },
    "mx": Object {
      "valid": true,
    },
    "regex": Object {
      "valid": true,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "valid": true,
    },
    "typo": Object {
      "valid": true,
    },
  },
}
`;

exports[`validation tests passes when subaddressing is allowed 1`] = `
Object {
  "valid": true,
  "validators": Object {
    "disposable": Object {
      "valid": true,
    },
    "mx": Object {
      "valid": true,
    },
    "regex": Object {
      "valid": true,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "valid": true,
    },
    "typo": Object {
      "valid": true,
    },
  },
}
`;

exports[`validation tests passes when valid special char 1`] = `
Object {
  "valid": true,
  "validators": Object {
    "disposable": Object {
      "valid": true,
    },
    "mx": Object {
      "valid": true,
    },
    "regex": Object {
      "valid": true,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "valid": true,
    },
    "typo": Object {
      "valid": true,
    },
  },
}
`;

exports[`validation tests passes when valid wildcard 1`] = `
Object {
  "valid": true,
  "validators": Object {
    "disposable": Object {
      "valid": true,
    },
    "mx": Object {
      "valid": true,
    },
    "regex": Object {
      "valid": true,
    },
    "smtp": Object {
      "valid": true,
    },
    "typo": Object {
      "valid": true,
    },
  },
}
`;

exports[`validation tests passes when we skip smtp validation 1`] = `
Object {
  "valid": true,
  "validators": Object {
    "disposable": Object {
      "valid": true,
    },
    "mx": Object {
      "valid": true,
    },
    "regex": Object {
      "valid": true,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "valid": true,
    },
    "typo": Object {
      "valid": true,
    },
  },
}
`;

exports[`validation tests passes with custom TLD 1`] = `
Object {
  "valid": true,
  "validators": Object {
    "disposable": Object {
      "valid": true,
    },
    "mx": Object {
      "valid": true,
    },
    "regex": Object {
      "valid": true,
    },
    "sanitization": Object {
      "valid": true,
    },
    "smtp": Object {
      "valid": true,
    },
    "typo": Object {
      "valid": true,
    },
  },
}
`;
